### PR TITLE
Address the 32k arguments bindings query limit

### PR DESCRIPTION
**Pull request by se7entyse7en**
Closes [DEV-484](https://athenianco.atlassian.net/browse/DEV-484).

### CODE REVIEW

**Comment by se7entyse7en** [comment]
@vmarkovtsev since you'll be available again on Monday I'd rather wait for you for this PR.

Anyway, I was able to locally both run the heater and render the dashboard with ES data. 👍
I first made it by manually checking the length on each query [here](https://github.com/athenianco/athenian-api/pull/533/commits/df22b5ba69f4f8ff03c8a5c87a275be8610827f1), then I made it smarter by applying it to everything and reverted that previous commit [here](https://github.com/athenianco/athenian-api/pull/533/commits/cc8d25c754bdbd33be833bf03c80aa7a296f3757).

File: server/athenian/api/models/__init__.py [hunk 1]
```diff
@@ -78,6 +80,48 @@ def create_base() -> BaseType:
...
+        return super(SmartColumn, self).in_(other)
+
+    def notin_(self, other):
+        """Change the `notin` operator into a `!= all` if too many values"""
+        log = logging.getLogger("smart_column.notin_")
+        log.info("len(other) = %d", len(other))
+        if len(other) > self.IN_AUTO_ANY_THRESHOLD:
+            log.warning("switching to `!= all`")
+            # `notin_` works if you pass a `dict` for example, but `all_` does not
+            return self != all_(other.__iter__())
```
Comment (se7entyse7en) [inline server/athenian/api/models/__init__.py:118]: I initially used `not_(self == any_(other))`, but it seems that `SQLAlchemy` is too smart... https://stackoverflow.com/questions/62952771/sqlalchemy-how-to-negate-an-any-expression

File: server/athenian/api/models/__init__.py [hunk 2]
```diff
@@ -78,6 +80,42 @@ def create_base() -> BaseType:
...
+
+    # 32767 is the maximum value for postgres:
+    #   - https://sentry.io/organizations/athenianco/issues/1785993425/
+    MAX_ALLOWED_QUERY_ARGUMENTS = 32767
+    # Anyway we cannot use exactly the upper limit because the length of the values used for
+    # `IN` and `NOT IN` could be near the the maximum allowed, but it often comes with other
+    # clauses that could make the total number of arguments bindings overflow the limit.
+    # This switch also affects performances:
+    #   - https://blog.jooq.org/2017/03/30/sql-in-predicate-with-in-list-or-with-array-which-is-faster/ # noqa: E501
+    IN_AUTO_ANY_THRESHOLD = 16383  # This specific value doesn't have any specific meaning
```
Comment (se7entyse7en) [inline server/athenian/api/models/__init__.py:94]: I'd like to also run all the metrics test with this value overridden to `0`, wdyt?

File: server/athenian/api/hacks/heat_cache.py [hunk 3]
```diff
@@ -99,6 +99,7 @@ async def async_run():
             if progress.finished_date is None:
                 log.warning("Skipped account %d / reposet %d because the progress is not 100%",
                             reposet.owner_id, reposet.id)
+                continue
```
Comment (vmarkovtsev) [inline server/athenian/api/hacks/heat_cache.py:102]: 🤦‍♂️ 🤦‍♂️ 🤦‍♂️

File: server/athenian/api/models/__init__.py [hunk 4]
```diff
@@ -78,6 +80,42 @@ def create_base() -> BaseType:
     return declarative_base(cls=(Refreshable, Explodable))
 
 
+class SmartColumn(Column):
+    """Column that automatically converts `IN` and `NOT IN` to `ANY` and `!= ALL` repsecitvely"""
+
+    # 32767 is the maximum value for postgres:
+    #   - https://sentry.io/organizations/athenianco/issues/1785993425/
+    MAX_ALLOWED_QUERY_ARGUMENTS = 32767
```
Comment (vmarkovtsev) [inline server/athenian/api/models/__init__.py:88]: Let's set it to 100 straight away. A big number of IN args results really bad performance that I saw in many profiles.

File: server/athenian/api/models/__init__.py (hunk 2 above)
Comment (vmarkovtsev) [reply server/athenian/api/models/__init__.py:94]: IMHO we don't need two thresholds as soon as we set the main one to 100.

File: server/athenian/api/models/__init__.py [hunk 5]
```diff
@@ -78,6 +80,42 @@ def create_base() -> BaseType:
     return declarative_base(cls=(Refreshable, Explodable))
 
 
+class SmartColumn(Column):
```
Comment (vmarkovtsev) [inline server/athenian/api/models/__init__.py:83]: Let's name it `AutoInAnyColumn`

**Comment by se7entyse7en** [comment]
done

**Comment by se7entyse7en** [comment]
Damn sqlite. 🤦 

@vmarkovtsev could you remind me why it's still relevant to test with sqlite? How we could circumvent this? Maybe the easiest is to change the threshold for sqlite tests?

**Comment by vmarkovtsev** [comment]
We continue to support SQLite because it is the best for running the unit tests locally. Sometimes a test crash, the DB cleanup does not run or is poorly written, and you end up with a screwed Postgres which you have to clean by hand or even reload the dump.

**Comment by vmarkovtsev** [comment]
I think we can do this: instead of coding the switch between IN and ANY inside `def in_`, it is possible to define a custom SQLAlchemy op with that switch inside and return it instead. That op can derive the threshold, but, unlike `Column`, when it renders it knows the dialect so it can override based on sqlite

**Comment by vmarkovtsev** [comment]
@se7entyse7en I am taking over this PR. I'll fix SQLite one way or another and merge.

**Comment by se7entyse7en** [comment]
> @se7entyse7en I am taking over this PR. I'll fix SQLite one way or another and merge.

okay thanks!

**Comment by se7entyse7en** [comment]
> when it renders it knows the dialect so it can override based on sqlite

Indeed that's what I missed when I was thinking on how to solve this 👍  Cool!

**Comment by vmarkovtsev** [comment]
Tests pass, but the new unit test does not.

**Pull request merged**